### PR TITLE
Add Timeouts functionality

### PR DESCRIPTION
**Pull request by italolelis**
## What does this PR do?

- Improves the Makefile
- Removes the `-race` flag from tests as this breaks on `alpine`, [this](https://github.com/golang/go/issues/14481) is golang issue
- Removes the `enable_load_balancing` property which was not used
- Removes the `ReadOnly` configuration for the admin API
- Removes the `Debug` configuration from the configuration as this was not used
- Makes the transport layer less smart and leave the stats tracking to the stats middleware
- Moves the balancer logic to its own package
- Adds an alias for the `roundrobin` load balancing which is called `rr` 
- Adds request and server timeouts and solves #320. Now we'll be able to configure timeouts per endpoits as well as timeouts for Janus server. Everything comes with defaults so this should not affect productions servers at HelloFresh. The new configurations are:

For Janus server:

```toml
#[respondingTimeouts]
# readTimeout is the maximum duration for reading the entire request, including the body.
#
# Optional
# Default: "0s"
#
# readTimeout = "5s"

# writeTimeout is the maximum duration before timing out writes of the response.
#
# Optional
# Default: "0s"
#
# writeTimeout = "5s"

# idleTimeout is the maximum duration an idle (keep-alive) connection will remain idle before closing itself.
#
# Optional
# Default: "180s"
#
# idleTimeout = "360s"
```

Per endpoint:

```json
{
"forwarding_timeouts": {"dial_timeout": "30s", "response_header_timeout": "60s"} 
}
```

### CODE REVIEW

File: pkg/proxy/balancer/rr.go [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+		mu      sync.RWMutex
+	}
+)
+
+// NewRoundrobinBalancer creates a new instance of Roundrobin
+func NewRoundrobinBalancer() *RoundrobinBalancer {
+	return &RoundrobinBalancer{}
+}
+
+// Elect backend using roundrobin strategy
```
Comment (vgarvardt) [inline pkg/proxy/balancer/rr.go]: I fixed this comment in the PR that was just merged =)

File: pkg/proxy/balancer/weight.go [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+	// ErrZeroWeight is used when there a zero value weight was given
+	ErrZeroWeight = errors.New("invalid backend, weight 0 given")
+)
+
+// NewWeightBalancer creates a new instance of Roundrobin
+func NewWeightBalancer() *WeightBalancer {
+	return &WeightBalancer{}
+}
+
+// Elect backend using roundrobin strategy
```
Comment (vgarvardt) [inline pkg/proxy/balancer/weight.go]: I fixed this comment in the PR that was just merged - `roundrobin` -> `weight` =)

File: pkg/proxy/balancer/weight.go [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+	WeightBalancer struct{}
+)
+
+var (
+	// ErrZeroWeight is used when there a zero value weight was given
+	ErrZeroWeight = errors.New("invalid backend, weight 0 given")
+)
+
+// NewWeightBalancer creates a new instance of Roundrobin
+func NewWeightBalancer() *WeightBalancer {
```
Comment (vgarvardt) [inline pkg/proxy/balancer/weight.go]: `Roundrobin` -> `WeightBalancer`

File: pkg/proxy/balancer/weight.go [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+	totalWeight := 0
+	for _, host := range hosts {
+		totalWeight += host.Weight
+	}
+
+	if totalWeight <= 0 {
+		return nil, ErrZeroWeight
+	}
+
+	r := rand.Intn(totalWeight)
```
Comment (vgarvardt) [inline pkg/proxy/balancer/weight.go:38]: I think we should add something like this to init rand seed

```
func init() {
	rand.Seed(time.Now().UnixNano())
} 
```

File: pkg/proxy/balancer/weight.go [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+	// WeightBalancer balancer
+	WeightBalancer struct{}
+)
+
+var (
+	// ErrZeroWeight is used when there a zero value weight was given
+	ErrZeroWeight = errors.New("invalid backend, weight 0 given")
+)
+
+// NewWeightBalancer creates a new instance of Roundrobin
```
Comment (vgarvardt) [inline pkg/proxy/balancer/weight.go]: `Roundrobin` -> `WeightBalancer`

**Comment by vgarvardt** [comment]
@italolelis can you pls split this PR into 2 or 3 to make changes more granular and review a bit easier? 

E.g. extract `-race`, `enable_load_balancing` and `Debug` related changes into the first PR, balancing changes/refactoring into another and Timeout functionality into the third one.

File: pkg/proxy/balancer/weight.go (hunk 4 above)
Comment (italolelis) [reply pkg/proxy/balancer/weight.go:38]: It's being initialized on the balancer.go file. It's for the whole package

**Comment by italolelis** [comment]
@vgarvardt The timeouts and balancer transport refactors have to be on the same PR, they are dependent. I could split the race stuff but they are tiny, that's why they are in the same PR. If you want I can split the `-race`, `enable_load_balancing` and `Debug` but not the other two

File: docs/config/proxy.md [hunk 6]
```diff
@@ -4,8 +4,9 @@
...
 | preserve_hosts        | Enable the [preserve host](/docs/proxy/preserve_host_property.md) definition           |
 | listen_path           | Defines the [endpoint](/docs/proxy/request_uri.md) that will be exposed in Janus       |
-| upstreams             | Defines the OAuth provider's [endpoints](/docs/proxy/upstreams.md) for the request   |
+| upstreams             | Defines the [endpoints](/docs/proxy/upstreams.md) that the request will be forwarded to|
 | strip_path            | Enable the [strip URI](/docs/proxy/strip_uri_property.md) rule on this proxy           |
-| enable_load_balancing | Enable load balancing for this proxy                                                   |
 | methods               | Defines which [methods](/docs/proxy/request_http_method.md) are enabled for this proxy |
 | hosts                 | Defines which [hosts](/docs/proxy/request_http_header.md) are enabled for this proxy   |
+| forwarding_timeouts.dial_timeout | The amount of time to wait until a connection to a backend server can be established. Defaults to 30 seconds. If zero, no timeout exists |
```
Comment (vgarvardt) [inline docs/config/proxy.md]: Can you point here that interval is in parsable `time.Duration` add link to Parse function with examples in go docs? Now the format is not clear here and in the parameter bellow.

File: pkg/config/specification.go [hunk 7]
```diff
@@ -13,7 +13,6 @@ import (
 // Specification for basic configurations
 type Specification struct {
 	Port                 int           `envconfig:"PORT"`
-	Debug                bool          `envconfig:"DEBUG"`
```
Comment (vgarvardt) [inline pkg/config/specification.go:16]: 👍

File: pkg/config/specification.go [hunk 8]
```diff
@@ -132,6 +138,8 @@ func init() {
 	viper.SetDefault("backendFlushInterval", "20ms")
 	viper.SetDefault("requestID", true)
 
+	viper.SetDefault("respondingTimeouts.IdleTimeout", 180*time.Second)
```
Comment (vgarvardt) [inline pkg/config/specification.go:141]: should not it be `respondingTimeouts.idleTimeout`? `idle` not capital - test is the best way to check it =)

Comment (italolelis) [reply pkg/config/specification.go:141]: Both work, but I'll add a test for it anyways

**Review by vgarvardt — APPROVED** [review]
ltgm

File: docs/config/proxy.md [hunk 9]
```diff
@@ -4,8 +4,9 @@
...
 | listen_path           | Defines the [endpoint](/docs/proxy/request_uri.md) that will be exposed in Janus       |
-| upstreams             | Defines the OAuth provider's [endpoints](/docs/proxy/upstreams.md) for the request   |
+| upstreams             | Defines the [endpoints](/docs/proxy/upstreams.md) that the request will be forwarded to|
 | strip_path            | Enable the [strip URI](/docs/proxy/strip_uri_property.md) rule on this proxy           |
-| enable_load_balancing | Enable load balancing for this proxy                                                   |
 | methods               | Defines which [methods](/docs/proxy/request_http_method.md) are enabled for this proxy |
 | hosts                 | Defines which [hosts](/docs/proxy/request_http_header.md) are enabled for this proxy   |
+| forwarding_timeouts.dial_timeout | The amount of time to wait until a connection to a backend server can be established. Defaults to 30 seconds. If zero, no timeout exists. You must use any format that is compatible with [time.Duration](https://golang.org/pkg/time/#Duration) |
```
Comment (rafaeljesus) [inline docs/config/proxy.md:11]: We might have side effects in some HF proxies I guess, can we make it 60s like it was before? 😢

File: pkg/config/specification.go [hunk 10]
```diff
@@ -26,17 +25,24 @@ type Specification struct {
...
 	Cluster              Cluster
+	RespondingTimeouts   RespondingTimeouts
 }
 
 // Cluster represents the cluster configuration
 type Cluster struct {
 	UpdateFrequency time.Duration `envconfig:"BACKEND_UPDATE_FREQUENCY"`
 }
 
+// RespondingTimeouts contains timeout configurations for incoming requests to the Traefik instance.
```
Comment (rafaeljesus) [inline pkg/config/specification.go]: `Traefik` typo

File: docs/config/proxy.md (hunk 9 above)
Comment (vgarvardt) [reply docs/config/proxy.md:11]: I think we should provide a "migration" script in upgrade docs to fill proxies with old default values.

Comment (rafaeljesus) [reply docs/config/proxy.md:11]: > Dial was always 30s

ah that's fine then

**Review by rafaeljesus — APPROVED** [review]

**Pull request merged**